server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -341,18 +341,23 @@ private void executeSearch(SearchTask task, SearchTimeProvider timeProvider, Sea
         final DiscoveryNodes nodes = clusterState.nodes();
         BiFunction<String, String, Transport.Connection> connectionLookup = buildConnectionLookup(searchRequest.getLocalClusterAlias(),
             nodes::get, remoteConnections, searchTransportService::getConnection);
+        assert nodeCount > 0 || shardIterators.size() == 0;
```

Comment:
if you have a condition in such an assert we should have a message with the actual values we don't know which one triggered.